[Optimization] Rewrite some code logic and method change for `post_meta.updated_at` option.

jackey8616 opened the pull request:
Accidentally close previous PR, sorry..

## PR Checklist
**Please check if your PR fulfills the following requirements:**

- [x] The commit message follows [our guidelines](https://github.com/theme-next/hexo-theme-next/blob/master/.github/CONTRIBUTING.md).
- [x] Tests for the changes have been added (for bug fixes / features).
   - [x] Muse | Mist have been tested.
   - [x] Pisces | Gemini have been tested.
- [ ] Docs have been added / updated (for bug fixes / features).

## PR Type
**What kind of change does this PR introduce?**  <!-- (Check one with "x") -->

- [x] Bugfix.
- [ ] Feature.
- [x] Code style update (formatting, local variables).
- [ ] Refactoring (no functional changes, no api changes).
- [ ] Build related changes.
- [ ] CI related changes.
- [ ] Documentation content changes.
- [ ] Other... Please describe:

## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->
With NexT _config.yml setting:
```yml  
post_meta:
  ...
  date_time_merge: true
  ...
```  
In every post 
ex create time is 2018-01-01 08:00:00 edite time is 2018-01-01 10:00:00
> Posted on 2018-01-01 08:00:00 | Post modified: 2018-01-01 10:00:00

But with this _config.yml setting:  
```yml  
date_time_merge: false
```  
updated_diff would not check by date and time but only date.
> Posted on 2018-01-01 08:00:00

## What is the new behavior?
By changing check statement from date to datetime.
_config.yml no longer need date_time_merge field, default will be check by also date and time.  
with default new filed in _config.yml setting:
```yml  
post_meta:
  ...
  # If set this option, the updated_diff will check diff base on this format.
  # Leave empty will check by both date and time.
  updated_diff_check_format:
  # Leave empty will use default hexo date_formt + ' ' + time_format for every post.
  datetime_format:
  ...
```  
> Posted on 2018-01-01 08:00:00 | Post modified: 2018-01-01 10:00:00
```yml  
updated_diff_check_format: YYYY-MM-DD HH
datetime_format:
```  
> Posted on 2018-01-01 08:00:00 | Post modified: 2018-01-01 10:00:00

```yml  
updated_diff_check_format: YYYY-MM-DD HH
datetime_format: YYYY-MM-DD HH a
```  
> Posted on 2018-01-01 08 am| Post modified: 2018-01-01 10 am

* Link to demo site with this changes: [MyGitHubPage again... lol](https://jackey8616.github.io/)

### How to use?
The default setting is empty, so the check and render all using full format.  
And this may effect some users who want to only resent YYYY-MM-DD  
By setting datetime_format to YYYY-MM-DD than everything won't be effect but only checking format.

```yml  
post_meta:
  ...
  # If set this option, the updated_diff will check diff base on this format.
  # Leave empty will check by both date and time.
  updated_diff_check_format:
  # Leave empty will use default hexo date_formt + ' ' + time_format for every post.
  datetime_format:
  ...
```  

and yes, it works!

## Does this PR introduce a breaking change?
- [ ] Yes.
- [x] No.

ivan-nginx commented:
As from user point of view if just turn on `updated_diff: true`:

```yml
# Post meta display settings
post_meta:
  item_text: true
  created_at: true
  updated_at: true
  categories: true
  # Only show 'updated' if different from 'created'.
  updated_diff: true
  # If set this option, the updated_diff will check diff base on this format.
  # Leave empty will check by both date and time.
  updated_diff_check_format:
  # Leave empty will use default hexo date_formt + ' ' + time_format for every post.
  datetime_format:
```

And see this:

![image](https://user-images.githubusercontent.com/16944225/38672797-f3d33e8e-3e57-11e8-8463-f55917d1260f.png)

> Post modified: 2017-10-03 07:02:31 

I think user don't want to see time (07:02:31) if dates already have different (2017-09-20 vs 2017-10-03).

***

1. Need to saw time's different's only if dates have no diffs.
2. Need to add description in configs. As for user, i don't understand what they meaning. I just turn on `updated_diff` option and don't understand what i can do with outher 2 options below. Maybe need to add descriptions with examples in config?
3. Keep it simple. As simplest as u can.

jackey8616 commented:
Totally agree with the idea about ' keeping simple'.

With time display style, here is some idea that comes out of my mind.
1. When create and updated date is same, but time is different, display date & time in create but only time in update.
ex created 2018-01-01 08:00:00 updated 2018-01-01 10:00:00
> Posted on 2018-01-01 08:00:00 | Post modified 10:00:00
2. When create and updated date is different, display only date.
ex created 2018-01-01 08:00:00 updated 2018-01-02 10:00:00
> Posted on 2018-01-01 | Post modified 2018-01-02
3. Left custom format is still work but write 1. and 2. just display it.
but 3. one is need to be more concern about it.
or maybe just cancel the field provide to user and let people use 1. or 2.
  
After all, the final _config.yml will look like:
```yml  
post_meta:
  ...
  updated_at:
    enable: false
    # Only show 'updated' if different from 'created'.
    updated_diff: false
    # If set this option, the updated_diff will check diff base on this format.
    # Leave empty will check by both date and time.
    diff_check_format:
  ...
```  
or maybe it have other features needs more filed.

ivan-nginx commented:
> or you think we should keep customize setting.

1. I think date and time format can be defined in Hexo, not in theme. So, no need this setting.
2. About first suggest don't understand where u want set it (in main or in title).

I think u can make commits and we will see to the way of the road how to look's that. Anyway, this is important option and no need to do this fast. Release of 6.2.0 will in April 15, so, we have a lot of time to improve this feature.

jackey8616 commented:
Oh, i have already committed,
maybe you can see on [my site](https://jackey8616.github.io/).

i added a under line at created and modified time for more user notice, 
and hover to it. the popping bubble will shows detail time.

jackey8616 commented:
i have removed the very begin's commit content about `datetime_format:`
this commit is just for showing under line and popping bubble  :laughing: 

later i will working on reducing digits.

ivan-nginx commented:
> later i will working on reducing digits.

Yeah, need to keep maximum web space, i know. Keep it maximum simpler.

jackey8616 commented:
Saw your ADDED,
About popping bubble.
Only showing **Posted on** and **Modified at** text at title, 
OR
Showing **Posted on (datetime)** / **Modified at (datetime)** ?
i only sure about 
hover onto it shows bubble with full datetime format(hexo full_date() function with hexo format).
i got bad understanding ... sorry. :sob: 

and quiet not understand with:
> Posted on 2018-01-01 — if same DAYS — and in this title — Post created: 2018-01-01 08:00 | Post modified: 2018-01-01 10:00.
what ** in this title** means?

if same DAYS, output both date and time like this ?
`Posted on 2018-01-01 08:00 | modified at 2018-01-01 10:00`

ivan-nginx commented:
> Showing Posted on (datetime) / Modified at (datetime) ?

Yeah, just like that.

> if same DAYS, output both date and time like this ?

If same days, no need output «Modified». Only `Posted on 2018-01-01` and then mouse over on label `2018-01-01`, **show in title** `Posted on (datetime) / Modified at (datetime)`.

I think any post created can be modified many times in one day, for example. So, it is some stupid to show «Modified» every time on every post **with** `updated_diff` option enabled. This option was created for this, for not show «Modified» label for one-day-created posts, and if we will show in each post «Modified» label, this option does not make sense.

jackey8616 commented:
i got you!
In every post only shows created time, 
and if the post have updated time,
they all put into popping bubble, is that right?

**show in title means HTML time tag's title attribute** :rofl: 

DIFF DAYS  => show both create / update time in post meat
SAME DAYS => show only create time in post meta, but updated time in bubble.

i'm also wondering every time i saved a .md that modified time will be change,
is kind on annoying..

ivan-nginx commented:
Title = Bubble, yes! It also called popup.

> i'm also wondering every time i saved a .md that modified time will be change,
is kind on annoying..

It make Hexo, we can't do anything here with this. We only can display it or not.

jackey8616 commented:
@ivan-nginx i have already make some changes on showing datetime,
and also removed _config.yml filed which this PR added at very first beginning.

you can also find some preview on my [site](https://jackey8616.github.io)
There are two post shows the changes under SAME DAYS and DIFF DAYS.

and here is something needs more opinion about:
When SAME DAYS, in the popup, the text seems a little duplicate on word 'POST'
and maybe we can add another text called modify: Modified
ex created 2018-01-01 08:00:00 update 2018-01-01 10:00:00
shows
> created 2018-01-01 08:00:00 / modified 10:00:00

ivan-nginx commented:
Yeah, seem's good.
I'll check code a little later.

ivan-nginx commented on source/css/_common/components/post/post-meta.styl:
```diff
@@ -16,6 +16,10 @@
   }
+
+  time {
+    border-bottom: 1px solid $grey-dark;
```
MB no need this styles?

jackey8616 replied:
i was thinking MB a underline would helping user's to notice that there is more info about time.
not underline but MB something else would not confuse with href link.

ivan-nginx replied:
Yeah, need to think about it, but underline seems like it's hyperlink and really do some confuse. :smiley:

jackey8616 replied:
MB we can replace solid to dotted?

ivan-nginx replied:
Try it.

ivan-nginx replied:
Or better dotted?
Try to play around with this: https://www.w3schools.com/cssref/pr_border-bottom_style.asp

jackey8616 replied:
I use dashed.
[Done! and take a look!](https://jackey8616.github.io)

jackey8616 commented:
@ivan-nginx  i have already push the latest stuff about underline style and popup.
take a look!
[DIFF_DAY_SAME_TIME and SAME_DAY_SAME_TIME](https://jackey8616.github.io)
[SAME_DAY_DIFF_TIME](https://jackey8616.github.io/page/2/)

ivan-nginx commented:
Yeah! For now seem's pefect. I'll edit it a little.

ivan-nginx commented:
Alright, i test it and there is some bugs: if i turn off `updated_diff` option or turn it on — there is nothing changes.

I suggest what with `updated_diff` option enabled `updated` date must appear always as `created` date, even if only time is change (like in [TEST-DIFF-DAY-SAME-TIME](https://jackey8616.github.io/)). But with turned `updated_diff` option on — it is fine for now, as we can see [here](https://jackey8616.github.io/page/2/).

ivan-nginx commented:
* Replaced replaced `only_diff` by `another_day` option. For now config is:

    ```yml
    post_meta:
      item_text: true
      created_at: true
      updated_at:
        enabled: true
        another_day: true
      categories: true
    ```

* `another_day` option:
   1. If `true`, show updated date label only if `updated date` different from 'created date' (post edited in another day than was created). And if post will edited in same day as created, edited time will show in popup title under created time label.
   2. If `false` show anyway, but if post edited in same day, show only edited time.

ivan-nginx commented on languages/en.yml:
```diff
@@ -22,15 +22,16 @@ sidebar:
+  edited: Edited on
+  created: Created
+  modified: Modified
+  edit: Edit this post
```
I think no need long labels here. When user move mouse under the date, he understand what he see date/time of post creation/modification.
So, `Created 2018-04-14 12:51:10 / Modified 23:00:00` — look's fine i think.

sli1989 commented:
one thing seems strange.

![1](https://user-images.githubusercontent.com/8521181/40104009-bf744ea0-5921-11e8-83d4-ecf77c98024e.png)

it was updated in 2018-5-12, but today it's 2018-5-16, but shows time only today.
maybe need to add compare condition with the present time .
